src/framebuffer/swtfb_client.rs

```diff
@@ -0,0 +1,224 @@
+//! This implements the IPC part of a rM2Framebuffer client to interact with this server:
+//! https://github.com/ddvk/remarkable2-framebuffer
+//!
+//! The client is developed according to the spec here:
+//! https://github.com/ddvk/remarkable2-framebuffer/issues/11
+
+use super::mxcfb::mxcfb_update_data;
+use crate::framebuffer::screeninfo::{FixScreeninfo, VarScreeninfo};
+use memmap2::{MmapOptions, MmapRaw};
+use std::ffi::{c_void, CString};
+use std::fs::{File, OpenOptions};
+use std::io::Error as IoError;
+use std::os::unix::prelude::AsRawFd;
+use std::{env, mem, ptr};
+
+const SWTFB_MESSAGE_QUEUE_ID: i32 = 0x2257c;
+
+pub const WIDTH: i32 = 1404;
+pub const HEIGHT: i32 = 1872;
+
+#[allow(non_upper_case_globals)]
+pub const maxWidth: i32 = 1404;
+#[allow(non_upper_case_globals)]
+pub const maxHeight: i32 = 1872;
+pub const BUF_SIZE: i32 = maxWidth * maxHeight * std::mem::size_of::<u16>() as i32; // hardcoded size of display mem for rM2
+const SEM_WAIT_TIMEOUT: i32 = 200000000; /* 200 * 1000 * 1000, e.g. 200ms */
+
+/// long on 32 bit is 4 bytes as well!!
+#[derive(Debug, Clone, Copy)]
+#[repr(i32)]
+#[allow(non_camel_case_types)]
+#[allow(dead_code)]
+pub enum MSG_TYPE {
+    INIT_t = 1,
+    UPDATE_t = 2,
+    XO_t = 3,
+    WAIT_t = 4,
+}
+
+#[derive(Debug, Clone, Copy)]
+#[allow(non_camel_case_types, dead_code)]
+pub struct xochitl_data {
+    pub x1: i32,
+    pub y1: i32,
+    pub x2: i32,
+    pub y2: i32,
+
+    pub waveform: i32,
+    pub flags: i32,
+}
+
+#[derive(Debug, Clone, Copy)]
+#[allow(non_camel_case_types, dead_code)]
+pub struct wait_sem_data {
+    /// C string
+    pub sem_name: [u8; 512],
+}
+
+/// MSG_TYPE has to match swtfb_update_data !!!
+#[derive(Clone, Copy)]
+#[allow(non_camel_case_types, dead_code)]
+pub struct swtfb_update {
+    pub mtype: MSG_TYPE,
+    pub data: swtfb_update_data,
+    //ms: u64,
+}
+
+#[derive(Clone, Copy)]
+#[allow(non_camel_case_types, dead_code)]
+pub union swtfb_update_data {
+    pub xochitl_update: xochitl_data,
+    pub update: mxcfb_update_data,
+    pub wait_update: wait_sem_data,
+}
+
+pub struct SwtfbClient {
+    msqid: i32,
+    do_wait_ioctl: bool,
+}
+
+impl Default for SwtfbClient {
+    fn default() -> Self {
+        let msqid = unsafe {
+            libc::msgget(
+                SWTFB_MESSAGE_QUEUE_ID,
+                libc::IPC_CREAT | libc::SHM_R | libc::SHM_W,
+            )
+        };
+        assert!(msqid >= 0);
+
+        // Does nested need special handling?
+        // This may not be needed at all.
+        if env::var("RM2FB_ACTIVE").is_ok() {
+            env::set_var("RM2FB_NESTED", "1");
+        } else {
+            env::set_var("RM2FB_ACTIVE", "1");
+        }
```

Comment:
No idea if it's needed. How does one can produce the nested case?
```cpp
// from ddvk's ipc.cpp : nesting only triggers logging a line
  uint16_t *mem =
      (uint16_t *)mmap(NULL, BUF_SIZE, PROT_WRITE, MAP_SHARED, fd, 0);

  if (getenv("RM2FB_NESTED") == "") {
    fprintf(stderr, "OPENED SHARED MEM: /dev/shm%s at %x, errno: %i\n",
            name.c_str(), mem, errno);
  }
  return mem;
```

Reply:
Not sure either. Seems like a debugging case but the behaviour isn't strictly altered when a nested app is detected. The [spec/design](https://github.com/ddvk/remarkable2-framebuffer/issues/4) also doesn't mention anything here.

Reply:
Is handling this important in the client @raisjn ? Seems to be merely used for more accurate debug output.

Reply:
it's not necessary, mostly was about debugging some stuff. not even sure nesting works correctly. i think rmkit uses RM2FB env variable to detect if it is on rm2, but don't think libremarkable needs it

Reply:
Okay. Remove this with commit 6a8dbe16 for now then. If this is ever needed for advanced checking of errors or such, we can add it again.